### PR TITLE
Issue 136 and 156

### LINKED ISSUE #136: Use python-libzim
in place of zimwriterfs

### LINKED ISSUE #156: Upgrade to Python3.11 + update dependencies versions
While starting gutenberg2zim on my Mac for the first time in a fresh venv, I got the following error :
```
Traceback (most recent call last):
  File "/Users/benoit/Repos/gutenberg/gutenberg2zim", line 16, in <module>
    from gutenbergtozim.download import download_all_books
  File "/Users/benoit/Repos/gutenberg/gutenbergtozim/download.py", line 19, in <module>
    from gutenbergtozim.export import get_list_of_filtered_books, fname_for
  File "/Users/benoit/Repos/gutenberg/gutenbergtozim/export.py", line 18, in <module>
    from jinja2 import Environment, PackageLoader
  File "/Users/benoit/Repos/gutenberg/.venv/lib/python3.9/site-packages/jinja2/__init__.py", line 33, in <module>
    from jinja2.environment import Environment, Template
  File "/Users/benoit/Repos/gutenberg/.venv/lib/python3.9/site-packages/jinja2/environment.py", line 15, in <module>
    from jinja2 import nodes
  File "/Users/benoit/Repos/gutenberg/.venv/lib/python3.9/site-packages/jinja2/nodes.py", line 19, in <module>
    from jinja2.utils import Markup
  File "/Users/benoit/Repos/gutenberg/.venv/lib/python3.9/site-packages/jinja2/utils.py", line 624, in <module>
    from markupsafe import Markup, escape, soft_unicode
ImportError: cannot import name 'soft_unicode' from 'markupsafe' (/Users/benoit/Repos/gutenberg/.venv/lib/python3.9/site-packages/markupsafe/__init__.py)
```

Searching the Internet indicates that downgrading markupsage to 2.0.1 solves the issue
```
pip install markupsafe==2.0.1
```

I confirm it solves the issue on my side. Probably linked to the fact that Jinja2 version is pretty old (2017).

Some details :
```
Python 3.9.12
pip 22.3.1
MacOS Monterey 12.6
```

### PR DESCRIPTION
**Pull request by benoit74**
Fix #136 
Fix #156

### CODE REVIEW

**Comment by benoit74** [comment]
Some questions / remarks:
- some CLI flags have been removed because they do not makes anymore sense
- I did not achieved to find if it is still possible to create a ZIM without index with zimscraperlib
- I did not achieved to find the documentation regarding the standard minimal Python code guideline (black + isort + flake with 88 chars max length if I did it appropriately in iFixit)
- I did a smoke test with few random books, do you have any idea of which book IDs are particularly relevant to test because they represent a corner case?

**Comment by kelson42** [comment]
> * I did not achieved to find the documentation regarding the standard minimal Python code guideline (black + isort + flake with 88 chars max length if I did it appropriately in iFixit)

It’s here https://github.com/openzim/overview/wiki but clearly not complete.

**Review by rgaudin — CHANGES_REQUESTED** [review]
Thank you for this large contribution!

I've added a lot of inline comments ; Should all be easy fixes.

The code is not working at the moment though for two reasons:
- the language information passed to export_books that gets transformed for ZIM metadata
- links are broken ; pointing to out-of-bound level (tools.js)
- links also include namespace prefixes that should be removed.

Another issue that I believe have nothing to do with this PR is that when I request a list of formats (epub, html), I do have the proper formats in my ZIM but the UI shows a link to PDF as well because the Book supports PDF (in the DB) but… it's not included because I didn't wanted it.
Code seems to ignore this possibility. We'll open a ticket once this is merged.

FYI Here's the command I used:

```sh
python gutenberg2zim -z gut_fr_test.zim -b 18812 -l fr -f epub,html
```

**Comment by benoit74** [comment]
Thank you for all these comments, I will work on it later this week, seems
pretty easy to fix.

File: gutenbergtozim/database.py [hunk 1]
```diff
@@ -2,20 +2,14 @@
...
-)
+from peewee import Model  # SqliteDatabase,
+from peewee import BooleanField, CharField, ForeignKeyField, IntegerField, TextField
 from playhouse.apsw_ext import APSWDatabase
 
 from gutenbergtozim import logger
 
 # db = SqliteDatabase('gutenberg.db')
-timeout = 1000 * 60 * 5 * 16
+timeout = 60 * 5 * 16
```
[the remark this answers is not in the document]
Comment (benoit74) [reply gutenbergtozim/database.py:12]: No idea either why it is so high, it is the max indeed. So it caused an integer capacity overflow multiplied by 1000 ;-)
Is it ok for you if I set it to 10 secs ? (I do not see any reason to wait for a lock more than 10 secs, we are not supposed to place any long duration locks on the database with current operations)

[the remark this answers is not in the document]
Comment (rgaudin) [reply gutenbergtozim/database.py:17]: I can imagine some lower hardware taking time to process DB query but 10s looks reasonable. We know this exists so should that be an issue, we'll have an idea.

**Comment by benoit74** [comment]
@rgaudin, I fixed all inline comments except the one regarding the SQL timeout, see proposition above.

I also fixed following issues (well spotted 👏):
- the language information passed to export_books that gets transformed for ZIM metadata
- links also include namespace prefixes that should be removed

I don't achieve to reproduce this issue (could you provide more info?):
- links are broken ; pointing to out-of-bound level (tools.js)

Regarding the link to the PDF which should not be displayed, I confirm this seems not linked to current PR. Issue is present in many places: cover article (gutenbergtozim/templates/cover_article.html), HTML version of a book (gutenbergtozim/templates/book_infobox.html) and bookshelves (2 times in gutenbergtozim/templates/js/tools.js).

File: gutenbergtozim/database.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (benoit74) [reply gutenbergtozim/database.py:17]: Yep, and this is only a busy timeout, i.e. a timeout for queries which do not achieve to progress because it is waiting for a table which is locked (for modification). In other words, even if the busy timeout is set to 10 secs, you can have queries taking much longer.
Cf. https://sqlite.org/c3ref/busy_timeout.html

File: gutenberg2zim [hunk 2]
```diff
@@ -51,27 +51,27 @@ help = (
 -u --rdf-url=<url>              Alternative rdf-files.tar.bz2 URL
 -b --books=<ids>                Execute the processes for specific books, """
     """separated by commas, or dashes for intervals
--c --concurrency=<nb>           Number of concurrent process for download """
+-p --concurrency=<nb>           Number of concurrent process for download """
```
[the remark this answers is not in the document]
Comment (rgaudin) [reply gutenberg2zim:51]: this has not been answered

**Review by rgaudin — APPROVED** [review]
Sorry for the delay ; have been busy.

This looks good ; There's a couple minor stuff you could look at but this is finally ready. thank you.

**Comment by kelson42** [comment]
@benoit74 Any chance to fix the CodeFactor issues? Usually the only one which are tolerated are the "complex method" ones (which would need a massive revamping and often are older than the PR itself).

**Comment by rgaudin** [comment]
> @benoit74 Any chance to fix the CodeFactor issues? Usually the only one which are tolerated are the "complex method" ones (which would need a massive revamping and often are older than the PR itself).

@kelson42 the codefactor issues are from the legacy code. They show up now because the file was touched (to fix the links) but were not introduced now.

File: gutenberg2zim (hunk 2 above)
[the remark this answers is not in the document]
Comment (benoit74) [reply gutenberg2zim:51]: I just changed the documentation because it was wrong on this line, the correct parameter is `-p` (cf line 25)

[the remark this answers is not in the document]
Comment (benoit74) [reply gutenberg2zim:51]: I would have preferred to use `-c` since there is no conflict on this parameter + it is more straightforward, but it would be a breaking change.

File: gutenbergtozim/export.py [hunk 3]
```diff
@@ -151,15 +166,15 @@ def export_skeleton(
...
-        Global.add_item_for(
-            path=tpl_path,
-            content=rendered,
-            mimetype="text/html",
-            is_front=True,
-        )
+    tpl_path = "Home.html"
+    template = jinja_env.get_template(tpl_path)
+    rendered = template.render(**context)
+    Global.add_item_for(
```
[the remark this answers is not in the document]
Comment (benoit74) [reply gutenbergtozim/export.py:172]: I'm ok with this as well.

[the remark this answers is not in the document]
Comment (benoit74) [reply gutenbergtozim/export.py:172]: Oh no, I'm not, sorry, tpl_path is used twice, that's the reason why I did it like this (once to get the appropriate template, another one to add it to the Zim)

File: gutenbergtozim/shared.py [hunk 4]
```diff
@@ -38,7 +38,7 @@ def setup(filename, language, title, description, name):
         Global.creator = Creator(
             filename=filename,
             main_path="Home.html",
-            favicon_path="favicon.png",
+            favicon_path="illustration",
```
[the remark this answers is not in the document]
Comment (benoit74) [reply gutenbergtozim/shared.py]: Done

File: gutenberg2zim (hunk 2 above)
[the remark this answers is not in the document]
Comment (rgaudin) [reply gutenberg2zim:51]: Ok let's do it. I'll adapt zimfarm if necessary

**Comment by benoit74** [comment]
> > @benoit74 Any chance to fix the CodeFactor issues? Usually the only one which are tolerated are the "complex method" ones (which would need a massive revamping and often are older than the PR itself).
> 
> @kelson42 the codefactor issues are from the legacy code. They show up now because the file was touched (to fix the links) but were not introduced now.

Yep, I opened [an issue](https://github.com/openzim/gutenberg/issues/158) to fix this, I would clearly prefer to have this PR merged first. This PR has changed many parts of the code, so delaying it too much is taking a risk of merge conflicts later. And I'm clearly not an expert in JS code, and this tools.js script seems a bit scary to fix without spending some time to understand all its purpose and confirm it is still working as expected afterwards.

**Comment by benoit74** [comment]
> Good to merge ;I'll do it tomorrow; thanks

I can do it if you prefer, @kelson42 granted me sufficient rights. Do you mind if I squash all commits? Are you ok with fast-forward merges? (these are my preferences, but I clearly don't mind to follow other conventions)

**Comment by rgaudin** [comment]
Please do. Our convention is to use a merge commit though
